### DISCUSSION TITLE
Добавить аналог using из C#

### DISCUSSION THREAD

**Discussion by MrSpecular** [pascalabcnet/pascalabcnet#2466] [Ideas]
```
begin
  using (var instance: FinalizeableObject := new FinalizeableObject()) do
  begin
    // Actions
  end;
end.
```
[Оператор using (Справочник по C#)](https://docs.microsoft.com/ru-ru/dotnet/csharp/language-reference/keywords/using-statement)

Предлагаю оставить эту **Issue**, пометив её как **enhancement**.

**Comment by ibond84** [comment]
Вы привели какую-то смесь из C# и паскалевского кода. Не принимается.

**Comment by MrSpecular** [comment]
1) Приведите аргументы **против** этой смеси. Слова

> **Не** принимается.

меня не убеждают.
2) Разве **PascalABC.Net** не является смесью **лучшего** из языков **Pascal, C# и Haskell** (возможно, список больше)?
3) Почему бы при закрытии **Issue** не указывать причину? Мне, например, неясно почему Вы закрыли данный **Issue**.

**Comment by ibond84** [comment]
ну, синтаксис смешивать же не надо. = это равно в паскале, а не присвоить. типы описываются после переменной. придумайте что-то ближе к паскалю.

**Comment by MrSpecular** [comment]
Подправил. На мой взгляд, объемлющие скобочки будет правильно требовать от пользователя только в случаях, если создаются несколько объектов. Точнее, сделать следующие записи валидными:
Пример 1:
```
begin
  using var a: T := new T() do // Скобочки не требуются, так как создается один объект.
  begin
    // Actions
  end;
end.
```
Пример 2:
```
begin
  using (var a: T := new T(), var b: T1 := new T1()) do // Скобочки обязательны, создается несколько объектов.
  begin
    // Actions
  end;
end.
```
. Причина желания такого поведения - стандартизация кода. В **if** объемлющие скобочки **необязательны**, в **while** также.

**Comment by ibond84** [comment]
Ну, вот как сделано в оксигене

```
using fs := new FileStream('textfile.txt', FileMode.Open) do 
begin
  var b := new byte[123];
  fs.Read(b, 0, b.Length);
end; // the filestream is closed here.
```

**Comment by MrSpecular** [comment]
Вариант однозначно хороший. Предлагаю сделать как в **Oxygen**.

**Comment by SunSerega** [comment]
Кстати, в паскале есть уже ключевое слово `with`, почему его не использовать? У него ведь похожее значение. То есть
```
begin
  with a: T := new T() do
  begin
    //
  end;
end.
```

**Comment by MrSpecular** [comment]
@SunSerega, скорее всего потому, что в **C#** и **Oxygen** оно для этой цели **не** используется. С **with** другая ассоциация - **упрощение доступа к членам класса или записи**. Думаю, что если использовать **with**, то это может создать трудности при переходе на **C#** или **Oxygen** (в том плане, что будет **не** совсем привычно). Но с другой стороны, конструкция **with {something} do** читается вполне понятно. И в обучении это может сыграть **положительную** роль. Так как ниша языка - обучение, то я думаю, что можно сделать вариант с **with**.

**Comment by SunSerega** [comment]
Зато в паскале нет `using`. И всё же, у них похожие значения. В C# ведь using используется для подключения пространства имён ещё (то есть, для влития их в основное пространство имён). А в паскале работает следующее:
```
begin
  with System.IO do
  begin
    var f := new FileInfo('');
  end;
end.
```
(ну или [почти](https://github.com/pascalabcnet/pascalabcnet/issues/1137)).

**Comment by MrSpecular** [comment]
Почему у **Issue** нет метки **enhancement**?

**Comment by miks1965** [comment]
Да, конструкция однозначно хорошая. Она предназнаяена для автоматического освобождения разных ресурсов.

Вот как я вижу ее использование в Паскале:
```
using var f := OpenRead('a.txt') do
begin
  f.ReadlnString.Print;
  ...
end;
```

Только делать ее муторно

**Comment by SunSerega** [comment]
А почему, всё же, не `with`?

**Comment by devapromix** [comment]
Видимо потому, что тут происходит еще и освобождение ресов.

**Comment by miks1965** [comment]
Верно. Предвижу, что тут сейчас начнется бешеный тред за отстаивание with и о том, что у нас свой путь.

Прошу - поберегите слова для чего-то более высокого :)

**Comment by ibond84** [comment]
Уже не вижу смысла в этом, если можно вызывать Dispose

**Comment by miks1965** [comment]
Ну в смысле Close.

Там такая фишка - если происходит исключение в секции using, то файл всё равно закрывается. А f.Close в этом случае не сработает и файл останется висеть открытым. И надо будет try finally городить

**Comment by ibond84** [comment]
Нет, Dispose. Close просто закрывает файл

**Comment by MrSpecular** [comment]
Открыл [голосование](http://forum.mmcs.sfedu.ru/t/golosovanie-with-using-v-analoge-using-iz-c/2466) по поводу выбора **with/using**. Подобные вопросы лучше решать голосованиями.

**Comment by GlebSBrykin** [comment]
А касательно данной записи:
```
using WinForms = System.Windows.Forms;
Begin
End.
```
Переопределение имени неймспейса или модуля для текущего файла кода.

**Comment by DanielArgent** [comment]
Позвольте обобщить имеющиеся варианты.

**Необходимость данной конструкции.**
Мне кажется очевидным, что использование оператора `using` является более предпочтительным, нежели использование конструкции `try ... except ... end;` в сочетании с вызовом метода `IDisposable.Dispose()`.

**Варианты дизайна.**
Несмотря на желание многих людей использовать ключевое слово `with`, оно уже является зарезервированным для других целей. Использовав его мы можем пойти двумя путями:
1. Изменить его для новых целей, тем самым отказавшись от его прежнего назначения и потеряв обратную совместимость.
2. Добавить новую семантику к уже имеющейся, тем самым добавив неоднозначность, вероятно сломав обратную совместимость (ведь теперь любой `IDisposable` объект в операторе `with` будет неожиданно для пользователя освобождать ресурсы).

_Следовательно, использование with невозможно._

Я вижу 2 потенциально возможных ключевых слова: `using` и `try`. 

Использование `try` (практикующееся в Java 7 и выше) удобно тем, что нам не нужно вводить новое ключевое слово, оно подходит по смыслу и с ним конструкция может выглядеть так:
```
try var f := OpenRead('a.txt') do
begin
  ...
end;
```
Возможная трудность при таком подходе: дать понять парсеру что это не обычный механизм обработки исключений (то есть это не начало `try ... except ... end;`):
```
// RAII-конструкция
try var f := OpenRead('a.txt') do
  ...

// Обработка исключений
try 
  var f := OpenRead('a.txt')
  ...
```
Из примеров выше видно, что при парсинге следует обращать внимание на символ новой строки и ключевое слово `do`.

Использование using, как уже отмечалось выше, более знакомо программистам на C#. Оно явно помечается ещё не использованным ключевым словом, так что будет разбираться легче.
```
using var f := OpenRead('a.txt') do
begin
  ...
end;
```
Кроме того, у этой формы есть ещё одно преимущество. В дальнейшем можно сделать механизм, появившийся в C# 8: использование using до конца текущей области видимости
```
function myFunc(fileName: string);
begin
  using var f := OpenRead('a.txt');
  f.ReadlnString.Print;
  // f.Dispose() here
end;
```
Таким образом, наиболее рациональным кажется использование `using` в форме
```
using var <var-name> = <resource-initialization> do
  <body>
```

**Способ реализации.**
Реализовать можно, воспринимая выражение 
```
using var <var-name> = <resource-initialization> do
  <body>
```
аналогом
```
do
  var <var-name> = <resource-initialization>;
  try
    <body>
  finally
    <var-name>.Dispose();
  end;
end;
```
Очевидно, что если `<resource-initialization>` возвращает объект, не реализующий интерфейс `IDisposable`, должно генерироваться  исключение времени компиляции.
